cancellation.md

```diff
@@ -1879,6 +1934,89 @@ void grow(vector<int>& v){
+
+
+> "I’m not particularly attached to any of the keywords here. I’m only reusing “try” out of parsimony. Bjarne has said that “try”  wasn’t in the original design for exceptions, but just added because people felt uncomfortable introducing a block without a keyword to head it."
+> _Lisa Lippincott_
+
+A simplistic approximation of the behavior could be expressed with setjmp/longjmp if longjmp respected destructors:
+
+```cpp
+jmp_buf buf;
+const auto cancel_if_possible = [&]()
+    {
+    try
+        {
+        if ( condition ) longjmp( buf, 1 );
+        }
+    catch ( ... )
+        {}
+    };
+
+if ( !setjmp( buf ) )
+    {
+    // cancellable effort…
+    }
+else
+    {
+    // code to execute if the effort is cancelled
+    // due to this condition…
+    }
+```
+
+Here’s the detailed behavior: each cancel_if block associates a “cancellation condition” and a handler block with a cancellable effort. The lifetime of the cancellation condition is the duration of the cancellable effort. If more than one cancellation condition is associated with the same effort, the lifetime of an earlier-declared condition starts before and ends after the lifetime of a later-declared condition. [This ordering is arbitrary.]
+
+During some parts of its lifetime, a cancellation condition may be “suspended.” During its lifetime, when not suspended, a cancellation condition is “active.” A cancellation condition is active at the start of its lifetime.
+
+The cancel_if_possible statement evaluates the active cancellation conditions, from eldest to youngest. If evaluation of a condition produces a true result, no further conditions are evaluated, the stack is unwound to the end of the associated cancellable effort, and control is transferred to the corresponding handler.  [Eldest-to-youngest because when a larger effort may be cancelled that includes all the younger cancellable efforts.]  If evaluation of a cancellation condition exits with an exception, the exception is handled, and the condition is treated as if it had produced false. If no cancellation produces true, execution continues with the following statement.
+
+For each of the following events, each cancellation condition that is active at the start of the event is suspended for the duration of the event; after the event, such conditions are reactivated if their lifetime has not ended.
+- Stack unwinding
+- The evaluation of a cancellation condition
+- The execution of a noexcept function
+
+> [Here “noexcept” should be taken to mean “no early exit” — this will cause noexcept function to never be cancelled. This is preferred to introducing “nocancel” because the guarantee callers need doesn’t distinguish between exceptions and cancellation. This assumes that there is no use for guarantees of “no cancellation but maybe exceptions” or “no exceptions but maybe cancellation."  The name “noexcept” is unfortunate in this regard.]
```

Comment:
I do see a use for "no exceptions but maybe cancellation". There will be algorithms that are noexcept and can be cancelled. The simplest of those algorithms is called never(). never() does not call value and does not call error and will only call done when it receives the request_stop() signal. never() is used as a placeholder in expressions that are conditional (if the feature is enabled perform this operation otherwise never().

Would you like to make changes to this?